### DISCUSSION TITLE
Workflow for rapid and flexible postgres prototyping with frequent schema changes

### DISCUSSION THREAD

**Discussion by bpeck81** [supabase/supabase#11212] [General]
This is to share my learnings and help beginners get started with data modeling in postgres. I'm new to postgres and recently started my journey with Supabase. 

For the workflow I use now, skip to **Solution for rapid prototyping**.

### The problem
If you have a really simple application, you probably don't need many views or functions. But views are really good when you want to apply complex queries (ex. ranking algorithms, or personalized feed customizations), or on top of your tables before consuming on the front end. 
The problem is that when you start creating views and functions that reference other views and functions, you end up with a big problem: dependency hell. If you want to drop a column on a view and change it's schema that other views reference, postgres will force you to drop (cascade) all objects directly and indirectly dependent on it. In my case I had 30 direct or indirect dependencies on a single view, and had to drop all of them and recreate all of them anytime I wanted to update it's schema. 
In this way, **postgres is really, really rigid.**
This led me to look into "migrations", hoping there was something there that would address this.

### About Migrations
When you're creating a database, people usually use at least two databases that they try to keep in sync: development (dev) and production (prod). With supabase, dev is housed on your local machine (in their docker container image) and prod is on supabase's servers. Your workflow to change your prod database is supposed to be: 
1. Make all your changes to you dev database
2. Use a tool (supabase cli) to calculate the difference between your two databases
3. The tool outputs sql code describing how to "migrate"/change your prod database to match your dev
4. You save that code to a file called a "migration" file
5. You then push that file with the supabase cli, which is executed on the remote server and changes your prod database. 

Any time you make a change to your dev database, you generate a new migration file. When you push your migration file, supabase goes through the history of all the files you've ever made to create the final resulting schema of your database. These migration files are expected to stack up over time.
But after 2 days of experimenting with all of  this, hoping it would solve my problem, I realized:
1. _Migrations using diffs absolutely suck_. The diff tools are really unintelligent. They cant determine what order to apply changes they generate in. Unless you have a really simple database, I wouldn't bother with using this method despite supabase recommending it.
2. They have no idea how to deal with dropping objects with dependencies. They will generate code telling you to drop and recreate a view, but not handle if that view needs to be dropped with cascade, and rebuilding all the dependent objects.

Also note, there's nothing in migrations about helping you with object dependencies in the dev database you're prototyping in. If you still want to change an object (view) in your dev database, you have to drop cascade and restore all the dependencies.

### About dependency storing
To prototype in my dev database, I eventually went on a journey to create a set of functions that would store and restore the definitions in postgres for all the objects dependent on the object I was about to drop/replace. There were a few github repos I found that tried to handle storing and restoring dependencies but none of them worked (ie https://gist.github.com/mateuszwenus/11187288#file-save_restore_dependencies-sql). 
The work flow for these functions would be
1. Use a function to save the definitions used to create the dependencies on the object you want to update. Postgres has definitions in its library.
2. Drop the object (view) with cascade
3. Recreate the object
4. Restore the object dependencies using what was saved

I eventually made some that worked, but ran into a new issue! (Here's what I made but I don't recommend this method https://github.com/bpeck81/dep_postgres)
New problem: Postgres saves the definitions used to make your views and functions after they've been _compiled_ at the time you made them.
The stored definitions _are not_ the queries you necessarily wrote! They're the ones you wrote _compiled_. This means your query with `select *` changes to `select col1, col2...`. So when you want to update a view to have another column and follow the workflow above, you can restore all of the objects, but the ones that you originally wrote to `select *` from it won't propagate that new column through.
For clarity if I have,
`create view ViewB as select * from ViewA`
When I later decide I want to add a column to ViewA, and restore ViewB using the method above, ViewB will not receive the new column from ViewA. Again, this is because ViewB stored the columns in ViewA at the time ViewB was created, and it did not store *.

### About Liquibase and Flywheel
Some searching eventually takes you to Liquibase and Flywheel as the standards for updating databases. They use XML (kind of like HTML or JSX) to create commands on how to change your database and do migrations. 
They work similarly to the migration process above, creating a migration script when you want to make a change, except they abstract the sql code you would write into simple tags.
For example:
`create or replace view ViewA as .... tons of code ....`
becomes something like
`<create view ViewA  path="point to my definition"/>`

Suddenly you can do tons of operations to change you database with just a few lines of code. I've read people use structure a migration file to change a view by, dropping all views and functions at the top, making their change, and restore all views and functions at the bottom. They would need to restore the views in the correct order, but it works. These migration files execute on your prod database in a single "transaction", meaning as long as you rebuild everything you drop there will be no down time when this is executed on your database.

Using one of these and this method I think is the best method in production, but it kind of sucks for prototyping. 

### Solution for rapid prototyping
What if you want to build a project in a week and you're just starting out? I've opted to ignore all the extensive tooling above for now and modify the prod database directly. Building initially and for beginners should be easy, quick, fun, and flexible. After that long 3 day journey, I ended up realizing the best method to deal with dependency hell and super quickly prototype is to use an extremely simple set of functions I wrote and a new mental model for updating your db. The secret to the whole thing is we just have to save the queries we write in our database ourselves.
Here's the repo: https://github.com/bpeck81/dep_commit_postresql

The model is:
1. Anytime we define a new object, we store it's definition in a table in our database (this allows us to store `*` in our queries)
2. When we want to update a schema/definition we: drop cascade everything dependent on it,  grab the definition from our table, redefine it, resave the definition to our table, and restore _everything_ in the database. 

I created a set of functions and table which I pushed to a repo for others to use. The functions are:
```
dep_store -- the table where all your view and function definitions are stored. Schema is (name text, definition text)
dep_commit(name, definition) -- save a definition
dep_restore() - restore everything saved (O(n^2))
dep_def(name) - quickly grab the definition of an object
dep_drop(name) - delete a definition
```

Here's the workflow I use:
1. `select * from dep_store order by name` - This shows me everything in my database in a glance and I can quickly see the beginnings of the query definitions for my objects.

2. `drop view myview cascade` - When I want to make a schema change to it

3. `create or replace view myview as <all the cool changes I want to make to the schema>` - Redefine the object

4. `select dep_commit('myview','create or replace view myview as...')` - We have to commit our changes to use them to restore in the future if the object is dropped via cascade

5. `select dep_restore()` Attempt to restore every object in the db. If we used * in our definitions, that will propagate through the newly restored objects. 

6. There may be an error at this step if a dependency needed a column you removed when making your change. In that case we look in the error message to see which object couldn't be restored and:

7. `select dep_def('brokenview')`

8. Copy the definition and paste it in the editor

9. Make edits needed

10. `select dep_commit(...)` - don't forget to commit all changes

11. `select dep_restore()`

Eventually we successfully restore everything!

When you want to drop something in your database, you just drop it and have to remember to call `dep_drop('name')` so it doesn't get restored.

**Again, anytime something is changed except tables in my `database,` I commit it with `dep_commit`.**

**I hope this helps people get started. I'm new to postgres too, so I'm definitely open to feedback from people with more experience!**

* Note if you use strings in your queries and you need to commit them, you have to change ' to ''. One single quote becomes two single quotes. 
* If you already have views and functions in your db, you'll have to commit all of them to use this method.
* Up until now, I've used the supabase editor for creating tables and I don't commit their definitions.

### Extra patterns I've uncovered
1. Views are really good at defining schemas and functions are good for filtering. I try to avoid returning a specific definition of function and always try to `returns setof some_view`.
2. This means I do pretty much all of my joins in views before sending them to functions to filter the views. I rely on the postgres query planner/optimizer to put my filters in the right spots. (Not sure this is best practice or optimizes things but I'm rolling with it for now to keep code clean).
3. Use postgres's Jsonb type and functions to jsonify and consume objects/views into other objects/views. 
4. Consider you database columns as split into two types of use - reading and operating on (don't necessarily split them into separate views, just consider them separately). Some columns are designed to be read others are designed to be operated on/with. Sometimes columns are designed for both uses. When reading it should probably just be jsonified and sent (joined) to the next view to consume, when operated on it should be kept as a column.
5. I started using pgAdmin4. At first it looks like overkill, but it's good. You just connect your db and make edits to your db with the query too.
6. I don't save any of my queries anywhere, they're saved in the db and if I need them I can dump the table where they're stored to a csv and check that into git.
7. If you want to return a count you can use a view calculate it on the fly or a trigger to update a table when it would change. I opted to use views for now.

**Comment by davepar** [comment]
Thanks for sharing all of this. I've run into the same problems prototyping my project. I thought one of Supabase's strengths was that it was good for prototyping, but managing Postgres change is not easy. My experience:

* Defining view and functions in the UI works pretty good at first, but you can't track changes in source control among other issues.
* `create or replace` works fine for functions until you modify the params or return type. Then you have to drop and create.
* Dropping objects with dependencies is a major headache.
  * You can `drop cascade`, but then need to re-create the dropped items.
  * You can drop everything and re-define everything, but even this has to be done in the right order.
* There are a bunch of existing diff and migration tools out there, but they all have their limitations.
  * The Supabase migration seems to rely on running a local instance of the entire stack, which is not simple. Definitely a buzz kill for a hosted service.
  * I also noticed some code in their tool that is doing some very expensive steps (e.g. starting docker and creating/defining a database) just to calculate schema diffs.
  * In particular, any tool that relies on the definitions in the pg_* tables will run into issues because those definitions are compiled, e.g. `select *` is stored as `select <col1>, <col2>, etc`. (I didn't know about this, but you saved me a ton of wasted time going down this dead end.)

I like your approach, but I'd prefer storing my object definitions in files I can track with git. That way the code changes stay in sync with the DB definitions. My preferred dev cycle (and this could be used for migrations as well) would be:

* Edit one or more SQL files (not XML or some other weird format) that each contain an obj (table, view, function) definition.
* Run a command:
  *  It loops through all of the files and for each one that has changed:
    * If the object already exists:
      * Uses the pg_* tables to determine the dependencies for the object. (I think the dependencies are accurate).
      * Executes a `drop <obj> cascade`.
    * Creates the obj based on the SQL file definition.
  * Re-create all of the dropped objects from the SQL files.

That's obviously a tall order, but I think it's possible by cobbling together parts of existing tools (for diffs and a dep tree).

**Reply by bpeck81** [reply]
Yeah it's completely reasonable to want to track changes in git, which my solution obviously doesn't support. I was originally thinking of using my method on the local container db and rely on schema diffing and checking the migrations into git. I realized it doesn't work though since schema diffing is really bad with these kinds of changes. 

What you described is possible and similar to my solution, where just keeps the queries in files rather than on the db. It's also a little more efficient in restoring dependencies rather than everything (I got lazy and didn't code this optimization since it doesn't change big-O). It could probably be built in another day, but I eventually decided I didn't want to waste another day of work on a side problem (after wasting 3). I settled for a rough solution that let me work on a prototype fast, very comfortably, and without care of tracking history, but everyone has their own needs. I actually think it would be really smart if Supabase built a tool/extension asap that did all this for the developer. It's like such a gaping problem that comes with postgres, and Idk how it had never been discussed or addressed. There's practically nothing on the internet about it. 

My strategy is to just to use my method (it's actually really comfortable to use) until I've nailed down my schema to a large degree, then use a migration tool like liquibase once I need prod features and struggle a little more through schema changes. Developers/founders should be focused on establishing product-market fit, and not this stuff!

Glad I helped and happy to further collaborate :)

**Reply by davepar** [reply]
Agreed that the prototyping stage shouldn't be weighed down with a heavy process. And also agree that Supabase should invest in this area. Presenting the UI as a simple way to modify tables and add RLS and functions is great, but they need to back it up with some migration infrastructure that isn't based on tape and glue. Tracking database schema changes in Postgres is complicated. More complicated than I expected.

Happy to collaborate as well, but I think I'm going to put this on the back-burner for now. I've got a website to create. :)

**Reply by bpeck81** [reply]
Cool, jw did you end up using the functions or approach I made?

**Reply by davepar** [reply]
I haven't tried your functions yet. I started defining all of my policies, views, and functions in files with drop/create statements for each. Probably stick with that for now.

**Comment by cloudorbush** [comment]
Completely agreed on the terrible DX. As soon as you need to apply a minor change such as adding or removing a column and you have functions and views you're facing a nightmare of complexity. Some kind of git based versioning but for postgres is needed. I think planetscale have developed that though they use mysql.
